IDP Getting Started guide (#30019)
DataDog/documentation: content/en/getting_started/internal_developer_portal/_index.md

> ---

Reviewer: Note: Need to rename/update the menu file entry to use the correct URL and name. It is currently "Software Catalog" https://docs-staging.datadoghq.com/domalessi/idp-getting-started/getting_started/software_catalog/

      tag: 'Documentation'
      text: 'Engineering Reports'
    - link: '/internal_developer_portal/overview_pages'
      tag: 'Documentation'
      text: 'Overview Pages'
---

{{< img src="tracing/internal_developer_portal/scrolling_the_catalog.mp4" alt="A video that scrolls through the Internal Developer Portal Catalog page and clicks on a service to show a dependency graph with parent and child services represented" video=true >}}

## Overview

Datadog's Internal Developer Portal (IDP) helps you unify software metadata, live telemetry, and developer workflows in a single platform. This guide walks you through setting up each core IDP component:

- [Software Catalog][1]: Real-time inventory of entities and environments, enriched with ownership and operational metadata.
- [Scorecards][2]: Measure adoption of engineering best practices using rules-based evaluations.
- [Self-Service Actions][3]: Enable developers to execute standardized tasks with a single click.
- [Engineering Reports][4]: Visualize quality, reliability, and compliance metrics across your stack.
- [Overview Pages][5]: Give developers and teams a personalized view of their entities, issues, and action items.

> Whether you're starting from scratch or integrating with existing systems like Backstage or ServiceNow, this guide will help you get started with IDP.

Author: ```suggestion
Whether you're starting from scratch or integrating with existing systems like Backstage or ServiceNow, use this guide to get started with IDP.
```

## Prerequisites

If you have not already, create a [Datadog account][6]. 

## Step 1: Populate Software Catalog 

> IDP starts with a real-time [Software Catalog][1], which inventories the key building blocks of your software architecture. In Datadog, these are called entities—they can represent individual services, APIs, or grouped Systems.

Reviewer: ```suggestion
IDP starts with the [Software Catalog][1], a real-time inventory of your software architecture's key building blocks. In Datadog, these are called entities—they can represent individual services, APIs, or grouped Systems.
```

You can add entities from:

- [APM, USM, and RUM][7]: Automatically discovered based on application telemetry
- [Manual definitions][8]: Created through the [Datadog app][9] or imported via tools like Terraform, Datadog APIs, or the GitHub integration
- [Third-party systems][10]: Through integrations with ServiceNow and Backstage

> While a basic definition registers an entity, you can enrich entity definitions with metadata to provide context like:

Reviewer: ```suggestion
A basic definition registers an entity to the catalog. You can make it more useful by enriching it with metadata to provide more context, including:
```
Light suggestion for a slight re-word. The original structure tripped me up a bit.

- Ownership info and team contacts
- Documentation, dashboards, runbooks
- Deployment pipelines and config links
- Production readiness data through unified service tagging

> The following example represents an entity definition for a `shopping-cart` service from an e-commerce application. It includes metadata about the service's owning team, languages used, runbook link, and code repository.

Author: ```suggestion
The following example defines a `system` entity representing an application composed of multiple services. It includes metadata such as display name, ownership, contacts, related documentation, integrations, and associated service components.
```

{{< code-block lang="yaml" filename="entity.datadog.yaml" disable_copy="true" collapsible="true" >}}
apiVersion: v3
  kind: system
  metadata:
    name: myapp

>     displayName: My App

Reviewer: Would be worth re-verifying this example. Seems to me it should be ` kind: service` and the name should be `shopping-cart` based on the intro above?

    tags:
      - tag:value

>     links:
>       - name: shopping-cart runbook
>         type: runbook
>         url: https://runbook/shopping-cart
>       - name: shopping-cart architecture
>         provider: gdoc
>         url: https://google.drive/shopping-cart-architecture
>         type: doc

Author: ```suggestion
    links:
      - name: shopping-cart runbook
        type: runbook
        url: https://runbook/shopping-cart
      - name: shopping-cart architecture
        provider: gdoc
        url: https://google.drive/shopping-cart-architecture
        type: doc
      - name: shopping-cart Wiki
        provider: wiki
        url: https://wiki/shopping-cart
        type: doc
      - name: shopping-cart source code
        provider: github
        url: http://github/shopping-cart
        type: repo
```

    contacts:
      - name: Support Slack
        type: slack
        contact: https://www.slack.com/archives/shopping-cart
    owner: myteam
    additionalOwners:
      - name: opsTeam
        type: operator
  integrations:
    pagerduty:
      serviceURL: https://www.pagerduty.com/service-directory/Pshopping-cart
  spec:
    components:
      - service:myservice
      - service:otherservice
{{< /code-block >}}

Read the [Software Catalog setup guide][11] to learn how to add or import entities, and review the [entity model reference][12] for schema details.

## Step 2: Evaluate entity quality with Scorecards

Use Scorecards to assess whether entities meet your organization's standards. Scorecards can measure:
- Monitoring coverage
- Production readiness
- Security posture
- Internal tooling adoption
- Ownership and documentation

> Datadog Scorecards include 10 out-of-the-box rules across observability practices, ownership tagging, and production readiness checkpoints. You can group rules into levels to categorize them by their criticality-level 1 (basic expectations), level 2 (recommended practices), and level 3 (advanced/aspirational goals).

Reviewer: I think the list of criticality levels would be a bit easier to parse in a bulleted list, but it's also ok as is if you feel like the page has too many lists. 😄

In addition to using default rules, you can define custom rules to reflect your internal standards:

> You can also define custom rules:

Reviewer: ```suggestion
To define custom rules:
```
(Or just delete this sentence. Note: If you leave this, replace previous sentence's colon with a period.)

1. Go to the [Scorecards page][13] and click Create Rule.
1. Specify the rule name, the scorecard it belongs to, a description, and the owning team.
1. Send an outcome of `pass`, `fail`, or `skip` for each entity in one of the following ways:

>    - Manually through the Datadog app

Reviewer: ```suggestion
   - Manually through the Datadog UI
```
Vale seems to not like "Datadog app".

   - Programmatically through the Scorecards API
   - Automatically, using [Workflow Automation][14] to post outcomes on a schedule
1. View an overview of outcomes on the [Scorecards page][13].

Learn more about scorecard configuration and custom rules in the [Scorecards documentation][2].

## Step 3: Use Self-Service Actions

> Self-Service Actions let engineers run repeatable tasks through a UI or API. For example, use Self-Service Actions to:

Reviewer: ```suggestion
Self-Service Actions let you run repeatable tasks through a UI or API. For example, use Self-Service Actions to:
```

- [Create an S3 bucket with Terraform][16]
- [Scaffold a new project in GitHub][17]
- [Manage Kubernetes deployments][18]

> Actions can be backed by automation systems like Terraform, GitHub Actions, or internal scripts. There are over 1000+ pre-built integrations with tools across source code management (e.g. Github, Gitlab), ticketing and incident management (e.g. Jira, ServiceNow, Pagerduty), chat (e.g. Slack & Microsoft Teams), cloud providers (e.g. AWS, GCP, Azure), and more. You can connect with any additional endpoints, including private resources, leveraging HTTP requests and private action runners.

Reviewer: ```suggestion
Actions can be backed by automation systems like Terraform, GitHub Actions, or internal scripts. Self Service Actions offer over 1000+ pre-built integrations with tools across source code management (for example, GitHub and GitLab), ticketing and incident management (for example, Jira, ServiceNow, and PagerDuty), chat (for example, Slack and Microsoft Teams), cloud providers (for example, AWS, GCP, and Azure), and more. You can connect with any additional endpoints, including private resources, leveraging HTTP requests and private action runners.
```
Recommend making these capitalizations consistent and matching their branding. Also, are self-service action integrations separate from Agent-based integrations? If we are referring to Agent-based integrations, you can use `{{< translate key="integration_count" >}}` to use the most up-to-date value (which is 800+).

[5]: /internal_developer_portal/overview_pages
[6]: https://www.datadoghq.com/
[7]: /internal_developer_portal/software_catalog/set_up/discover_entities/
[8]: /internal_developer_portal/software_catalog/set_up/create_entities/
[9]: https://app.datadoghq.com/software
[10]: /internal_developer_portal/software_catalog/set_up/import_entities/
[11]: /internal_developer_portal/software_catalog/set_up/
[12]: /internal_developer_portal/software_catalog/entity_model/entity_types/
[13]: https://app.datadoghq.com/software/scorecards
[14]: https://app.datadoghq.com/workflow/blueprints?selected_category=SCORECARDS
[16]: https://app.datadoghq.com/app-builder/apps/edit?startModalOpen=false&template=create-new-s3-bucket&viewMode=templatePreview
[17]: https://app.datadoghq.com/app-builder/apps/edit?startModalOpen=false&template=scaffold-github&viewMode=templatePreview
[18]: https://app.datadoghq.com/app-builder/apps/edit?startModalOpen=false&template=manage-kubernetes-deployments&viewMode=templatePreview
[19]: https://app.datadoghq.com/software/self-service-actions
[20]: https://app.datadoghq.com/idp/reports
[21]: https://app.datadoghq.com/idp/overview/developer

> [22]: https://docs.datadoghq.com/internal_developer_portal/overview_pages/

Reviewer: ```suggestion
[22]: /internal_developer_portal/overview_pages/
```
